labrad/node/__init__.py

```diff
@@ -503,95 +506,91 @@ class NodeServer(LabradServer):
+        # stopped.
         self.instances = {}
-        self.initMessages(True)
+
+        # ServerProcess instances that will be notified when new servers connect
+        # to labrad. ServerProcesses use this information to determine when the
+        # subprocesses that they launch have successfully connected to labrad.
+        # We use a separate collection here than self.instances because
+        # ServerProcess instances need to receive the server connection
+        # notification in order to startup, but they are not added to
+        # self.instances until after they have started up successfully.
+        # TODO: ServerProcess instances should manage notifications on their
+        # own, since they have access to a labrad client.
+        self.instances_to_notify = weakref.WeakSet()
 
     @inlineCallbacks
     def initServer(self):
-        """Initialize this server."""
+        """Initialize this server after connecting to the labrad manager."""
         self.config = yield NodeConfig.create(self)
         self.refreshServers()
         self.autostart(None)
 
     def stopServer(self):
         """Stop this node by killing all subprocesses."""
-        self.initMessages(False)
         stoppages = [inst.stop() for inst in self.instances.values()]
         return defer.DeferredList(stoppages)
 
 
     # message handling
 
-    def initMessages(self, connect=True):
-        """Set up message dispatching."""
-        def f(receiver, signal):
-            try:
-                if connect:
-                    dispatcher.connect(receiver, signal)
-                else:
-                    dispatcher.disconnect(receiver, signal)
-            except dispatcher.DispatcherError as e:
-                method = 'connect' if connect else 'disconnect'
-                msg = ('Error while setting up message dispatching: '
-                       'method={}, receiver={}, signal={}.')
-                print(msg.format(method, receiver, signal), e)
-        # set up messages to be relayed out over LabRAD
-        messages = ['server_starting', 'server_started',
-                    'server_stopping', 'server_stopped',
-                    'status']
-        for message in messages:
-            f(self._relayMessage, message)
-        # set up message handlers for subprocess events
-        f(self.subprocessStarting, 'server_starting')
-        f(self.subprocessStarted, 'server_started')
-        f(self.subprocessStopping, 'server_stopping')
-        f(self.subprocessStopped, 'server_stopped')
-
-    def _relayMessage(self, signal, sender, **kw):
+    def on_subprocess_message(self, sender, message):
+        if message == 'server_stopped':
+            instance_name = sender.name
+            if instance_name in self.instances:
+                del self.instances[instance_name]
+        self._relayMessage(message, server=sender.server_name,
+                           instance=sender.name)
```

Comment:
`server_name` is the name that appears in the node config file for the server and in the node web interface list of available servers. `name` is the name of the running server instance. These are the same for "global" servers that can only run on one node at a time, but may be different for "local" servers that can run per node, e.g. the server name is "GPIB Bus" but the name of an instance might be "mynode GPIB Bus".

The variables should be changed, though, because on server configs we use "name" and "instance_name" but then on instances these change to "server_name" and "name", respectively, so yeah, will clean this up in a future PR.